File: Shows/php/stage/Stage/Proxy.php
<?php

class Proxy implements Subject
{
    private $realSubject;

    public function __construct(RealSubject $realSubject)
    {
        $this->realSubject = $realSubject;
    }


    public function request(): void
    {
        if ($this->checkAccess()) {
            $this->realSubject->request();
            $this->logAccess();
        }
    }
    private function checkAccess(): bool
    {
        // Some real checks should go here.
        echo "Proxy: Checking access prior to firing a real request.<br/>";
        return true;
    }
    private function logAccess(): void
    {
        echo "Proxy: Logging the time of request.<br/>";
    }
}

File: Language/CSharp/Structural/Proxy/Example2/Show/Show/Proxy.cs
using System;

namespace Show
{
    public class Proxy : Subject
    {
        private RealSubject _realSubject;
        
        public Proxy(RealSubject realSubject)
        {
            this._realSubject = realSubject;
        }
        
        public void Request()
        {
            if (this.CheckAccess())
            {
                this._realSubject.Request();
                this.LogAccess();
            }
        }
        public bool CheckAccess()
        {
            // Some real checks should go here.
            Console.WriteLine("Checking access prior to firing a real request.");
            return true;
        }
        
        public void LogAccess()
        {
            Console.WriteLine("Proxy: Logging the time of request.");
        }
    }
}